protocol/describegroups/describegroups.go

```diff
@@ -0,0 +1,105 @@
+package describegroups
+
+import (
+	"github.com/segmentio/kafka-go/protocol"
+)
+
+type State string
+
+const (
+	Dead               State = "Dead"
+	Stable             State = "Stable"
+	AwaitingSync       State = "AwaitingSync"
+	PreparingRebalance State = "PreparingRebalance"
+)
+
+func init() {
+	protocol.Register(&Request{}, &Response{})
+}
+
+type Request struct {
+	GroupIDs []string `kafka:"min=v0,max=v0"`
+}
+
+func (r *Request) ApiKey() protocol.ApiKey { return protocol.DescribeGroups }
+
+func (r *Request) Group() string {
+	if len(r.GroupIDs) != 1 {
+		return ""
+	}
+	return r.GroupIDs[0]
+}
+
+func (r *Request) Map(cluster protocol.Cluster) ([]protocol.Message, protocol.Reducer, error) {
+	requests := make([]protocol.Message, len(r.GroupIDs))
+	for i, group := range r.GroupIDs {
+		requests[i] = &Request{GroupIDs: []string{group}}
```

Comment:
To reduce heap allocations you could do two things here:

* Allocate all requests in a single slice:
```go
requests := make([]Request, len(r.GroupIDs))
messages := make([]protocol.Message, len(r.GroupIDs))

for i := range r.GroupIDs {
    requests[i] = ...
    messages[i] = &requests[i]
}

return messages, new(Response), nil
```

* Takes slices in the input `GroupIDs` instead of constructing new slices with `[]string{group}`:
```go
for i := range r.GroupIDs {
    requests[i] = Request{GroupIDs: r.GroupIDs[i : i+1 : i+1]} // 3-index form to cap the capacity
    ...
}
```